java/kotlin-extractor/src/main/kotlin/utils/TypeSubstitution.kt

```diff
@@ -215,27 +217,36 @@ private fun matchingTypeParameters(l: IrTypeParameter?, r: IrTypeParameter): Boo
 }
 
 // Returns true if type is C<T1, T2, ...> where C is declared `class C<T1, T2, ...> { ... }`
-fun isUnspecialised(paramsContainer: IrTypeParametersContainer, args: List<IrTypeArgument>): Boolean {
+fun isUnspecialised(paramsContainer: IrTypeParametersContainer, args: List<IrTypeArgument>, logger: Logger): Boolean {
+    return isUnspecialised(paramsContainer, args, logger, paramsContainer)
+}
+
+private fun isUnspecialised(paramsContainer: IrTypeParametersContainer, args: List<IrTypeArgument>, logger: Logger, origParamsContainer: IrTypeParametersContainer): Boolean {
     val unspecialisedHere = paramsContainer.typeParameters.zip(args).all { paramAndArg ->
         (paramAndArg.second as? IrTypeProjection)?.let {
             // Type arg refers to the class' own type parameter?
             it.variance == Variance.INVARIANT &&
-            matchingTypeParameters(it.type.classifierOrNull?.owner as? IrTypeParameter, paramAndArg.first)
+                    matchingTypeParameters(it.type.classifierOrNull?.owner as? IrTypeParameter, paramAndArg.first)
         } ?: false
     }
     val remainingArgs = args.drop(paramsContainer.typeParameters.size)
 
-    val parentClass = paramsContainer.parents.firstIsInstanceOrNull<IrClass>()
+    val parentTypeContainer = paramsContainer.parents.firstIsInstanceOrNull<IrTypeParametersContainer>()
 
     val parentUnspecialised = when {
         remainingArgs.isEmpty() -> true
-        parentClass == null -> false
-        else -> isUnspecialised(parentClass, remainingArgs)
+        parentTypeContainer == null -> {
+            if (remainingArgs.isNotEmpty()) {
```

Comment:
Isn't this always true given the previous branch?

Reply:
Thanks, good point.